### PR TITLE
AVRO-3049: Add checks to BinaryDecoder for bytes length

### PR DESCRIPTION
**Pull request by john-karp**
Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO/) issues and references them in the PR title. For example, "AVRO-1234: My Avro PR"
  - https://issues.apache.org/jira/browse/AVRO-3049
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

testStringMaxArraySize
testNegativeBytesLength
testBytesMaxArraySize
testBytesMaxLengthProperty

### Commits

- [X] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

**Review by RyanSkraba — APPROVED** [review]
Thanks for the contribution -- this makes sense to me.

File: lang/java/avro/src/main/java/org/apache/avro/io/BinaryDecoder.java [hunk 1]
```diff
@@ -87,15 +91,26 @@ void clearBuf() {
 
   /** protected constructor for child classes */
   protected BinaryDecoder() {
+    String o = System.getProperty(MAX_BYTES_LENGTH_PROPERTY);
```
Comment (Fokko) [inline lang/java/avro/src/main/java/org/apache/avro/io/BinaryDecoder.java:94]: Should we also call `super()` here?

**Review by Fokko — APPROVED** [review]
One small remark. Looks good apart from that!

**Comment by RyanSkraba** [comment]
Cherry-picked into [branch-1.10](https://github.com/apache/avro/commit/4f9f97002e4d3e78766536563baceb36d02aeb0d).  You should see this fix in the next release.  Thanks for taking this on!

**Pull request merged**